cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -2082,43 +2103,135 @@ TEST(Substrait, BasicPlanRoundTrippingEndToEnd) {
+      "add", {compute::field_ref(FieldPath({1})), compute::field_ref(FieldPath({2}))});
+  // Acero only outputs the expressions mentioned in the projection but Substrait
+  // outputs existing columns plus the expression.
+  // For validation purpose, a project expression with expected projection
+  // field plus existing fields is created for Acero.
+  // And for Substrait just the project expression with expected projection.
+  std::vector<compute::Expression> acero_project_fp_exprs = {
+      compute::field_ref(FieldPath({0})), compute::field_ref(FieldPath({1})),
+      compute::field_ref(FieldPath({2})), project_fp_expr};
+  std::vector<compute::Expression> substrait_project_fp_exprs = {project_fp_expr};
+
+  auto acero_declarations = compute::Declaration::Sequence(
+      {compute::Declaration(
+           {"scan", dataset::ScanNodeOptions{dataset, scan_options}, "s"}),
+       compute::Declaration(
+           {"project", compute::ProjectNodeOptions{acero_project_fp_exprs}, "p"})});
+
+  // adding the project expression field to schema
+  ASSERT_OK_AND_ASSIGN(auto project_schema,
+                       dummy_schema->AddField(dummy_schema->num_fields(),
+                                              field("add(shared,distinct)", int32())));
+
+  ASSERT_OK_AND_ASSIGN(
+      auto expected_table,
+      GetTableFromPlan(acero_declarations, exec_context, project_schema));
+
+  std::shared_ptr<ExtensionIdRegistry> sp_ext_id_reg = MakeExtensionIdRegistry();
+  ExtensionIdRegistry* ext_id_reg = sp_ext_id_reg.get();
+  ExtensionSet ext_set(ext_id_reg);
+
+  // declaration which Substrait would expect since the idea is to do the projection
+  // which would produce an equivalent result to the declaration created for Acero with
+  // additional project fields.
+  auto substrait_declarations = compute::Declaration::Sequence(
+      {compute::Declaration(
+           {"scan", dataset::ScanNodeOptions{dataset, scan_options}, "s"}),
+       compute::Declaration(
+           {"project", compute::ProjectNodeOptions{substrait_project_fp_exprs}, "p"})});
+
+  ASSERT_OK_AND_ASSIGN(auto serialized_plan,
+                       SerializePlan(substrait_declarations, &ext_set));
+
+  ASSERT_OK_AND_ASSIGN(
+      auto sink_decls,
+      DeserializePlans(
+          *serialized_plan, [] { return kNullConsumer; }, ext_id_reg, &ext_set));
+  // assert project declaration
+  const auto& roundtripped_project =
+      std::get<compute::Declaration>(sink_decls[0].inputs[0]);
+  // assert project declaration
+  // Note: the provided expressions for Substrait declaration only contains one
+  // expression, but substrait produces expressions for the existing number of fields plus
+  // provided expression. Since the output expressions from the deserialized relation
+  // contains fields which weren't used in the project expression.
+  AssertProjectRelation(roundtripped_project, acero_project_fp_exprs, project_schema);
```

Comment:
I think this is wrong.  However, if you use my fix regarding emit up above you will probably still have to play games here because you will end up with two project nodes (since we deserialize an emit as a dedicated project node).

Reply:
How about not adding an emit, do it as the Substrait suppose to do things. So Acero output would give only the provided project expressions based output columns. Substrait would give original columns + project expression based output columns. In the validation, we drop the original columns and just compare the rest with Acero output. I think it would be much neater and do what the ToProto suppose to do rather than changing things to fit the Acero output. What do you think?

Reply:
If you do not add an emit then I don't know how you create an equivalent Substrait plan.  If the two plans generate different output then this is not the purpose of `ToProto`.  The goal is to create a Substrait plan that has the same output as the Acero plan.

So, for example, if you have (in Acero):

SCAN(schema={"x": int32(), "y": int64()})
PROJECT({field_ref("x")})

There is no way to create an equivalent Substrait plan without an emit

Reply:
Fair enough. Let's try that route.